### DISCUSSION TITLE
Waiting 750ms between Bluetooth calls and other oddities

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#1400] [Q&A] [answered]
Pardon the vague title.

As part of the work to move from 2.7.3 to 3.x I am changing some of the structure around how we interact with Shiny.BluetoothLE. I have come across a couple of oddities that I have previously ignored.

Taking this as an example 

```
public async Task DisconnectAsync()
{
    System.Diagnostics.Debug.WriteLine("DeviceWrapper DisconnectAsync");
    this.peripheral.CancelConnection();

    // HACK Delay added because android bluetooth needs time between calls
    await Task.Delay(750);

    await this.peripheral.WhenDisconnected().Timeout(TimeSpan.FromSeconds(1)).Take(1);

    this.UnHookDeviceStatusChanged();
}
```

1. "_HACK Delay added because android bluetooth needs time between calls_" this seems odd. Is it really necessary? This app has been using Shiny since the beginning so maybe it was needed long ago.

2. Can you see what the point might be in the call _WhenDisconnected().Timeout(TimeSpan.FromSeconds(1)).Take(1);_? Seems pointless to me

Thanks

**Comment by aritchie** [comment, marked as answer]
1. Some older android flavours used to take a couple of seconds to disconnect and it happens async.  CancelConnection is also meant to stop future connection attempts which is why it isn't async in the first place.  However, the test should likely be:

```cs
var task = this.peripheral.WhenDisconnected().Timeout(TimeSpan.FromSeconds(5)).Take(1).ToTask();
this.peripheral.CancelConnection();

await task.ConfigureAwait(false);

```

2. I guess it is just making sure the disconnect event fires.  If the device is already connected, it will fire once on hook anyhow.
